Add support for customizing HTTP headers

kyokuping opened the pull request:
This PR adds support for custom HTTP headers, introducing a `--http-header` CLI flag to accept `key=value` pairs. These headers can also be specified within the `setting.pkl` file.

Closes #633

SPICE: https://github.com/apple/pkl-evolution/pull/24

kyokuping committed:
Add support for customizing HTTP headers

HT154 reviewed (commented):
Thanks for the contribution! 🎉

I have a few primary concerns here:
* Although rare, it is valid to send a specific header more than once in a request. When there are headers with the same name their order may also matter.
* How does your implementation handle "reserved" headers that users should not override like `Origin` or `User-Agent`?
* It seems less than ideal to send the same headers to all hosts, eg. if host A requires auth but host B does not, Pkl should not send a configured `Authorization` header to both hosts. I think a prefix matching scheme like the one used for rewrites is necessary here to ensure the proper security controls are in place.

To this end, I think the Pkl API for this needs to be something along these lines:
```pkl
headers: Mapping<HttpRewrite, Listing<Pair<String, String>>>
```

This complicates the CLI a bit, of course, but maybe something like this is reasonable?
```
--http-header <prefix>='<header name>: <header value>'
```

kyokuping committed:
Allow customizing headers for different URLs

HT154 commented:
Just a heads up, we want reviews from the entire core team because this feature affects the core language and has security implications. One member of the team is currently out on leave so it's going to be a little bit before we can fully review this, but we haven't forgotten about it!

kyokuping committed:
add `@Since` annotation in pkl option

Co-authored-by: Jen Basch <jbasch94@gmail.com>

kyokuping committed:
validate `header` syntax

stackoverflow reviewed (commented):
The design sounds sensible at first glance, but due to the complexity and security issues it may cause it may be a good idea to create a SPICE for it.
What do you think @bioball, @HT154?

bioball reviewed (commented):
Hey, thanks for the PR! I agree with @stackoverflow; it'd be good to have this discussed in a SPICE.

If you'd like to drive this feature, can you write one up? It should be submitted as PR to [pkl-evolution](https://github.com/apple/pkl-evolution).

Your SPICE should start by copy/pasting the [template](https://github.com/apple/pkl-evolution/blob/main/spice-templates/SPICE-0000-pkl-template.adoc). Your SPICE number would be the next number after the latest proposed SPICE see (https://github.com/apple/pkl-evolution/pulls)

kyokuping committed:
Rename HttpRewrite to HttpPrefix for clarity

kyokuping committed:
Add strict HTTP header validation to EvaluatorSettings

bioball commented:
@kyokuping was the close intentional?

kyokuping commented:
Sorry, accidentally closed the PR. 😢

bioball commented:
Gotcha, feel free to re-open a new PR!

kyokuping committed:
Update implementation based on the SPICE changes

- Introduce pattern-based URL matches
- Both accept a single and multiple values for header values
- Move header append logic to RequestRewritingClient

HT154 commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/commands/BaseOptions.kt:
```diff
@@ -285,6 +290,57 @@ class BaseOptions : OptionGroup() {
+                  ?: fail("Header '$header' is not in 'name:value' format.")
+              require(HEADER_NAME_REGEX.matcher(headerName).matches()) {
+                "HTTP header name '$headerName' has invalid syntax."
+              }
```
Unlike the project/settings method of setting headers, the CLI doesn't exclude specific headers names and prefixes. It would be good to make these consistent.

HT154 reviewed (commented):
Hey @kyokuping would you mind rebasing this to resolve the conflict(s)?

I have some pretty minor feedback, but from a technical standpoint this looks pretty good. I'd like to try and ship this in 0.32 in a few months, so it'd be great if this was ready for another set of eyes by the time the core team's all back together in a few weeks!

kyokuping committed:
Revert Copyright date change

kyokuping committed:
Update `@Since` annotations

kyokuping committed:
Rewrite lambdas as expressions

kyokuping committed:
Disallow using commas as header separator

kyokuping committed:
Implement strict header validation in CLI

kyokuping committed:
Use computeIfAbsent

Co-authored-by: Jen Basch <jbasch94@gmail.com>

kyokuping commented on pkl-core/src/main/java/org/pkl/core/util/IoUtils.java:
```diff
@@ -854,4 +884,41 @@ public static void validateRewriteRule(URI rewrite) {
+    if (!headerName.equals(headerName.toLowerCase())) {
+      throw new IllegalArgumentException(
+          "HTTP header '%s' should be all lowercase".formatted(headerName));
+    }
```
Hey, had any chance discussing around this?

bioball replied:
We just chatted on this; we've settled on: we don't think it's necessary to require any special rules for header names when configured using CLI flags.

stackoverflow commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/commands/BaseOptions.kt:
```diff
@@ -285,6 +290,37 @@ class BaseOptions : OptionGroup() {
+
+        try {
+          for ((stringPattern, header) in it) {
+            val headerRegex = Regex("""^(.+?):[ \t]*(.+)$""")
```
```suggestion
          val headerRegex = Regex("""^(.+?):[ \t]*(.+)$""")
          for ((stringPattern, header) in it) {
```
Wasteful to recompile the Regex on every iteration. Even better to move it to the companion object.

stackoverflow commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/commands/BaseOptions.kt:
```diff
@@ -93,6 +95,9 @@ class BaseOptions : OptionGroup() {
       }
     }
+
+    val HEADER_NAME_REGEX = Pattern.compile("^[a-zA-Z0-9!#$%&'*+-.^_`|~]+$")
```
This is duplicated verbatim from `IoUtils`. You can make it public there and use it here.

stackoverflow commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/commands/BaseOptions.kt:
```diff
@@ -93,6 +95,9 @@ class BaseOptions : OptionGroup() {
     }
+
+    val HEADER_NAME_REGEX = Pattern.compile("^[a-zA-Z0-9!#$%&'*+-.^_`|~]+$")
+    val HEADER_VALUE_REGEX = Pattern.compile("^[\\t\\u0020-\\u007E\\u0080-\\u00FF]*$")
```
Same here (duplicated from IoUtils).

stackoverflow commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/commands/BaseOptions.kt:
```diff
@@ -285,6 +290,37 @@ class BaseOptions : OptionGroup() {
       .multiple()
       .toMap()
 
+  val httpHeaders: List<PPair<Pattern, List<PPair<String, String>>>> by
```
This is more of a nit, but: `List<PPair<Pattern, List<PPair<String, String>>>>` is a quite generic type that doesn't convey much information. It would be better to have a properly-named record that stores the URL glob and the header name-value pairs.

HT154 replied:
I agree this type's gotten a bit out of hand and a record would be a welcome improvement. I'm not eager to block this PR more than necessary, so I'm fine leaving this as follow-up work.

stackoverflow commented on pkl-core/src/main/java/org/pkl/core/util/IoUtils.java:
```diff
@@ -52,6 +52,36 @@ public final class IoUtils {
+    "via"
+  };
+
+  private static final String[] reservedHeaderPrefixs = {"proxy-", "sec-", "access-control-"};
```
```suggestion
  private static final String[] reservedHeaderPrefixes = {"proxy-", "sec-", "access-control-"};
```

stackoverflow commented on pkl-core/src/main/java/org/pkl/core/util/IoUtils.java:
```diff
@@ -52,6 +52,36 @@ public final class IoUtils {
+    "accept-charset",
+    "accept-encoding",
+    "access-control-request-headers",
+    "access-control-request-method",
```
Do you need these here if you already have `"access-control-"` in the `reservedHeaderPrefixs`?

stackoverflow commented on pkl-core/src/main/java/org/pkl/core/util/IoUtils.java:
```diff
@@ -854,4 +884,41 @@ public static void validateRewriteRule(URI rewrite) {
+  }
+
+  public static void validateHeaderValue(String headerValue) {
+    if (headerValueLike.matcher(headerValue).matches()) {
```
```suggestion
    if (!headerValueLike.matcher(headerValue).matches()) {
```
I'm pretty sure your `headerValueLike` is a positive regex, so **not** matching it should throw an error.

kyokuping committed:
move headerRegex compilation out of loop

kyokuping committed:
remove unused regex

kyokuping committed:
fix typos

kyokuping committed:
update reservedHeaderNames list

kyokuping committed:
remove lowercase enforcement for header names in CLI

kyokuping committed:
apply spotless formatting

kyokuping committed:
fix inverted regex condition in IoUtils

Co-authored-by: Islon Scherer <islonscherer@gmail.com>

bioball commented on pkl-core/src/main/java/org/pkl/core/evaluatorSettings/PklEvaluatorSettings.java:
```diff
@@ -149,8 +156,37 @@ public record Http(@Nullable Proxy proxy, @Nullable Map<URI, URI> rewrites) {
+                      .toList();
+              parsedHeaderDefs.add(new Pair(urlPattern, pairs));
+            } catch (InvalidGlobPatternException e) {
+              throw new PklException(ErrorMessages.create("invalidUri", stringPattern));
```
```suggestion
              throw new PklException(ErrorMessages.create("invalidGlobPattern", stringPattern));
```

bioball commented on pkl-core/src/main/java/org/pkl/core/http/HttpClientBuilder.java:
```diff
@@ -110,6 +113,12 @@ public Builder addRewrite(URI sourcePrefix, URI targetPrefix) {
   }
 
+  @Override
+  public Builder setHeaders(List<Pair<Pattern, List<Pair<String, String>>>> headers) {
```
I think it would be better if this signature matched the in-language evaluator settings.

I think this should be something like:

```java
setHeaderRules(Map<String, Map<String, List<String>>> headerRules)
```

Where the top-level map key is the glob pattern.
This method can throw `IllegalArgumentException` if a glob pattern is invalid.

And, I think we should also have:

```java
addHeaderRule(String globPattern, Map<String, List<String>> headers)
```

It'd also be good to avoid using `Pair` here.

bioball commented on stdlib/EvaluatorSettings.pkl:
```diff
@@ -235,3 +242,51 @@ class ExternalReader {
+  "proxy-"
+  "sec-"
+  "access-control-"
+}
```
```suggestion
local const reservedHttpHeaderPrefixes = Set
  "proxy-",
  "sec-",
  "access-control-",
)
```

HT154 commented on stdlib/EvaluatorSettings.pkl:
```diff
@@ -128,6 +128,9 @@ local const hasNonEmptyHostname = (it: String) ->
 typealias HttpRewrite = String(startsWith(Regex("https?://")), endsWith("/"), hasNonEmptyHostname)
 
+@Since { version = "0.32.0" }
+typealias UrlPattern = String(endsWith(Regex("[/*]")))
```
I think this validation needs to be tightened:
```suggestion
typealias UrlPattern = String(endsWith(Regex(#"/\*{0,2}"#)))
```

The idea here is that attempts to match a prefix should always end on a path boundary (exact or prefix) to avoid accidentally matching unintended origins, e.g. pattern `*.example.com` matches `foo.example.company.com`, which could leak credentials.

bioball replied:
I actually think we don't need to tighten this. Unlike string prefix matching, glob expressions are unlikely to be accidentally misconfigured.

These two are clear enough, in my opinion:

`https://example.com**` -- match everything that starts with `example.com`
`https://example.com/**` --  match everything within the host `example.com`

I do wonder if we want a `external boolean isGlobPattern` on class String, so this can be:

```
String(isGlobPattern)
```

HT154 replied:
Hmm I still think it would be ideal to prevent prefix matching on hosts, but I do see what you mean. I like the idea of `isGlobPattern` to mirror `isRegex`.

HT154 commented on pkl-commons-cli/src/main/kotlin/org/pkl/commons/cli/commands/BaseOptions.kt:
```diff
@@ -285,6 +287,37 @@ class BaseOptions : OptionGroup() {
+            IoUtils.validateHeaderName(headerName)
+            IoUtils.validateHeaderValue(headerValue)
+            headersMap
+              .computeIfAbsent(stringPattern) { mutableListOf() }
```
The pattern here isn't validated by the same rule as the pattern values in `pkl:EvaluatorSettings`

HT154 reviewed (commented):
I think this is getting real close now. From my perspective, just a little bit to hash out with URI patterns.

bioball reviewed (approved):
This is in a good enough place to merge; there's some things that I plan on addressing in a follow-up PR!

HT154 reviewed (approved):
With @bioball on this one.

Thanks again for all the work on this, @kyokuping! We really appreciate the contribution.
